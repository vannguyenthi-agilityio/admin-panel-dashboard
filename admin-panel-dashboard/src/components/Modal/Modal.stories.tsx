// Libs
import type { Meta, StoryObj } from "@storybook/react";

// Components
import Modal from ".";

const meta: Meta<typeof Modal> = {
  title: "Components/Modal",
  tags: ["autodocs"],
  component: Modal,
  argTypes: {
    open: { description: "open modal type is boolean"},
    onClose: { description: "funct close modal type is void"},
    children: { description: "body modal type is ReactNode"},
    showIconClose: { description: "check to show close icon type is boolean"},
    labelButton: { description: "label text to show button in footer of modal type is string"},
    title: { description: "title of modal type is string"},
    className: { description: "custom class for modal type is string"},
    size: { description: "custom size padding, text size for modal with option type: 'xs' | 'sm' | 'md' | 'lg'"},
  },
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    children: <p className="py-4"> Modal body text goes here.</p>
  },
};
