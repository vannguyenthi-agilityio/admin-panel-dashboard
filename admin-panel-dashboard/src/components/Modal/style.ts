import { cva } from 'class-variance-authority';

export const modalStyles = cva(
  'font-primary font-light rounded-xl shadow p-6 w-full transition-all',
  {
    variants: {
      variant: {
        primary: ["bg-white", "text-dark"],
        secondary: ["bg-transparent", "text-darker"],
      },
      size: {
        xs: ["text-xs", "px-4", "py-3"],
        sm: ["text-sm", "px-6", "py-5"],
        md: ["text-md", "px-8", "py-6"],
        lg: ["text-lg", "px-2", "py-2"],
      },
    },
    defaultVariants: {
      variant: "primary",
      size: "sm",
    },
  }
);
