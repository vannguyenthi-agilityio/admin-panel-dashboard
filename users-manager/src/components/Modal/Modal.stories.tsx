import Storybook from '@storybook/react';
import { action } from '@storybook/addon-actions';

// Components
import { Modal } from './index';

export default {
  title: 'Components/Modal',
  component: Modal
} as Storybook.ComponentMeta<typeof Modal>;

const Template: Storybook.ComponentStory<typeof Modal> = (args) => (
  <Modal
    onClose={action('onCloseHandler')}
    onSubmit={action('onSubmitHandler')}
    {...args}
  />
);

export const ModalDefault = Template.bind({});
ModalDefault.args = {
  title: 'Add User',
  isOpen: true,
  size: 'default',
  description: '',
  submitButtonText: 'Submit',
  cancelButtonText: 'Cancel',
  isDisabledSubmit: false,
  isLoading: false,
  isCentered: true,
  variant: 'primary'
};
