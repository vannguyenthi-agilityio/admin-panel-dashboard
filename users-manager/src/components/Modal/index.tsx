import React from 'react';
import {
  Modal as ModalChakra,
  ModalProps as ModalPropsChakra,
  ModalOverlay,
  ModalContent,
  ModalHeader,
  ModalFooter,
  ModalBody,
  CloseButton,
  Box
} from '@chakra-ui/react';

// Components
import { Button } from '../Button';
import { Text } from '../Text';

export interface ModalProps extends ModalPropsChakra {
  title?: string;
  description?: string;
  children: React.ReactNode;
  size?: string;
  variant?: 'normal' | 'primary' | 'secondary';
  bgColor?: string;
  submitButtonText?: string;
  cancelButtonText?: string;
  isDisabledSubmit?: boolean;
  isDisabledOutsideClick?: boolean;
  isLoading?: boolean;
  onSubmit?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Modal = ({
  isOpen = false,
  onClose,
  title,
  description,
  returnFocusOnClose = false,
  isCentered = false,
  children,
  size = 'default',
  variant = 'primary',
  bgColor = 'default.light',
  submitButtonText,
  cancelButtonText,
  isDisabledSubmit = true,
  isDisabledOutsideClick = false,
  isLoading,
  onSubmit,
  ...props
}: ModalProps) => {
  const btnRef = React.useRef();

  const handleOnClose = () => {
    if (!isDisabledOutsideClick) {
      onClose();
    }
    return;
  };

  return (
    <>
      <ModalChakra
        size={size}
        variant={variant}
        isOpen={isOpen}
        onClose={handleOnClose}
        finalFocusRef={btnRef}
        isCentered={isCentered}
        returnFocusOnClose={returnFocusOnClose}
        {...props}
      >
        <ModalOverlay />
        <ModalContent bg={bgColor}>
          <ModalHeader
            display="flex"
            justifyContent="space-between"
            alignItems="center"
          >
            <Text
              size="medium"
              variant="caption"
              letterSpacing="wide"
              w="100%"
              marginRight="30px"
              value={title}
            />
            <CloseButton
              outline="0"
              borderColor="transparent"
              color="default.grey.500"
              onClick={onClose}
            />
          </ModalHeader>
          <ModalBody>
            {description && <Text color="default.light" value={description} />}
            {children}
          </ModalBody>
          <ModalFooter pt="0px" display="flex" justifyContent="center">
            <Box>
              {submitButtonText && (
                <Button
                  isLoading={isLoading}
                  isDisabled={isDisabledSubmit}
                  label={submitButtonText}
                  onClick={onSubmit}
                  size="default"
                />
              )}
              {cancelButtonText && (
                <Button
                  label={cancelButtonText}
                  variant="transparent"
                  mr={3}
                  onClick={onClose}
                />
              )}
            </Box>
          </ModalFooter>
        </ModalContent>
      </ModalChakra>
    </>
  );
};
