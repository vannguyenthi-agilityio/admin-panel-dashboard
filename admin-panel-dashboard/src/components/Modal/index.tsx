import { memo, ReactNode } from "react";

// Helpers
import { clsxMerge } from '@/helpers';

// Components
import { Button } from "@/components";
import { CloseIcon } from "@/components/Icons";

// Styles
import { modalStyles } from './style';

interface IModal {
  open?: boolean;
  onClose?: () => void;
  children: ReactNode;
  showIconClose?: boolean;
  labelButton?: string;
  title?:string;
  className?: string;
  size?: 'xs' | 'sm' | 'md' | 'lg';
}

const Modal = ({
  open,
  onClose,
  children,
  showIconClose = false,
  labelButton = "",
  title = "Modal title",
  className,
  size,
}: IModal) => (
    <div onClick={onClose}
      className={`fixed inset-0 flex min-h-[100px] justify-center items-center transition-colors ${open ? "visible bg-black/20" : "invisible"}`}
    >
    {/* modal */}
    <div
      onClick={(e) => e.stopPropagation()}
      className={clsxMerge(
        modalStyles({
          size
        }),
        `${open ? "scale-100 opacity-100" : "scale-125 opacity-0"}`,
        className
      )}
      >
      {/* <!--Modal title--> */}
      <div className="flex flex-shrink-0 flex-wrap items-center justify-start rounded-t-md border-b-2 mb-2 border-neutral-100 border-opacity-100 py-4">
        <h5 className="text-xl font-medium leading-normal text-neutral-800 dark:text-neutral-200">
          {title}
        </h5>
      </div>
      {/* <!--Close button--> */}
      { showIconClose &&
        <Button
          onClick={onClose}
          className="absolute top-5 right-5 w-[40px] h-[40px] p-0 rounded-lg text-gray-400 bg-white hover:bg-gray-50 hover:text-gray-600"
          leftIcon={<CloseIcon stroke="black"/>}
        />
      }
      {/* <!--Modal body--> */}
      {children}
      {/* <!--Modal footer--> */}
      <div className="flex flex-shrink-0 gap-3 flex-wrap items-center justify-end rounded-b-md border-t-2 mt-2 border-neutral-100 border-opacity-100 pt-4">
        <Button
          onClick={onClose}
          label='Cancel'
        />
        {labelButton &&
          <Button
            onClick={onClose}
            size='small'
            label={labelButton}
            buttonType='blood'
          />
        }
      </div>
    </div>
  </div>
);

export default memo(Modal);
