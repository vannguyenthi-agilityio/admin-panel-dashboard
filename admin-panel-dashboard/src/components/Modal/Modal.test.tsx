import {
  render
} from '@testing-library/react';
import '@testing-library/jest-dom';
import Modal from '.'

describe('Modal component', () => {
  it('Renders modal component', () => {
    const {container} = render(<Modal children={<p className="py-4"> Modal body text goes here.</p>} />)
    expect(container).toMatchSnapshot();
  });
})
