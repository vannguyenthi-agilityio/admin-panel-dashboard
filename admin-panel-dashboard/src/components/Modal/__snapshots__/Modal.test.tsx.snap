// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component Renders modal component 1`] = `
<div>
  <div
    class="fixed inset-0 z-50 flex min-h-[100px] justify-center items-center transition-colors invisible"
  >
    <div
      class="font-primary font-light rounded-xl shadow p-6 w-full transition-all bg-white text-dark text-sm px-6 py-5 scale-125 opacity-0"
    >
      <div
        class="flex flex-shrink-0 flex-wrap items-center justify-start rounded-t-md border-b-2 mb-2 border-neutral-100 border-opacity-100 py-4"
      >
        <h5
          class="text-xl font-medium leading-normal text-neutral-800 dark:text-neutral-200"
        >
          Modal title
        </h5>
      </div>
      <p
        class="py-4"
      >
         Modal body text goes here.
      </p>
      <div
        class="flex flex-shrink-0 gap-3 flex-wrap items-center justify-end rounded-b-md border-t-2 mt-2 border-neutral-100 border-opacity-100 pt-4"
      >
        <button
          class="flex flex-row gap-x-4 font-primary uppercase font-medium disabled:cursor-not-allowed items-center justify-center bg-white text-darker hover:bg-greyish border hover:border-greyish border-greyish hover:shadow-md text-sm py-2 px-4 rounded-lg"
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
